### DISCUSSION TITLE
What are the advantages of using State vs Extensions?

### DISCUSSION THREAD

**Discussion by thomasmost** [tokio-rs/axum#1830] [General]
Specifically, a route handler for me currently looks something like this:

```rs
pub async fn listed(Extension(pool): Extension<Pool>) -> axum::Json<Vec<Org>> {
  let dao = &mut OrgDao::new(&pool);
  let orgs = dao.get_listed();
  axum::Json(orgs)
}
```

vs

```rs
pub async fn listed(State(state): State<AppState>) -> axum::Json<Vec<Org>> {
  let pool = &state.pool;
  let dao = &mut OrgDao::new(pool);
  let orgs = dao.get_listed();
  axum::Json(orgs)
}
```

...but what are the advantages of using State for db pool management/why should I? I understand that state is shared between all routes but I think that's still true with this initialization, yes?

```rs

#[tokio::main]
async fn main() {
  ...
  let pool = init_pool();
  let app = Router::new()
    .route("/", get(service_info))
    .nest("/v0", v0_api)
    .layer(Extension(pool));
  ...
}
```

**Comment by davidpdrsn** [comment]
`State` is type safe, `Extension` is not. If you forget `.layer(Extension(pool));` everything will still compile but fail at runtime. Not so with `State`.

**Comment by thomasmost** [comment]
Can you apply more than one state object to one router?

**Reply by davidpdrsn** [reply]
With sub states sorta yes https://docs.rs/axum/0.6.10/axum/extract/struct.State.html#substates

**Reply by thomasmost** [reply]
I'm late following up on this but **nice!** This is exactly what I was looking for. Seems like a great pattern that affords the right flexibility/simplicity I'm looking for

**Comment by Geo-W** [comment]
Is it possible to use both? 😂

**Reply by thomasmost** [reply]
Yes, in fact I frequently do (using Extensions for auth and State for things like the db pool that are more global)

```rust
pub async fn some_endpoint(
  State(pool): State<Pool>,
  Extension(SessionUser(user)): Extension<SessionUser>,
  fields: Json<PostBody>,
) -> http::StatusCode {
  ...
}
```

**Reply by Geo-W** [reply]
🙏 Thanks!

**Reply by mektigboy** [reply]
@thomasmost may I ask why? What are the advantages of using both? Why Extensions for auth and State for things like the db pool?

**Reply by thomasmost** [reply]
so with the disclaimer that I am not an Axum expert and there may be ways of adapting State to serve the same purpose, I'll say that my understanding is that State, while more type-safe than Extensions, must be guaranteed and applicable across any given router, whereas Extensions can be loaded dynamically for a subset of routes.

Therefore, to use State for authentication/authorization in a multi-tenant system with variable user roles, you would need to have a separate router for each type of user to load a role-scoped user in State. 

Instead I gate admin routes with an 'admin' middleware, which builds upon an 'employee' middleware and a 'session user' middleware (for a logged in user not necessarily a member of any tenant organization). An admin-gated route can therefore expect the common `ApplicationState` object with the database connection pool, and an `OrgAdmin` extension with the authorized user details.

See below for an example Router demonstrating what I mean:

```rs
  let api_organization = Router::new()
    // UPDATE a tenant organization details, requires a user with an ADMIN membership within the tenant org
    .route("/", put(routes_v1::org::update))
    .route_layer(axum::middleware::from_fn(middleware::admin::auth))
    // GET a tenant org details, requires a user with a valid membership within the tenant org
    .route("/", get(routes_v1::org::details))
    .route_layer(axum::middleware::from_fn_with_state(
      state.clone(),
      middleware::employee::auth,
    ))
    // CREATE a tenant org, requires a session user without any specific org membership
    .route("/", post(routes_v1::org::create))
    .route_layer(axum::middleware::from_fn_with_state(
      state.clone(),
      middleware::session_user::auth,
    ))
```

I'd be very happy to be informed of a better/more conventional way of doing this if there is one — @davidpdrsn do you have thoughts on this pattern?

**Reply by mektigboy** [reply]
Thanks for the great explanation!

**Reply by oscar6echo** [reply]
I see that your router uses a state without the applying `.with_state()`, implying that the latter is not necessary with `from_fn_with_state()`.   
Do I understand well ?  What if you still add `with_state()` ? What would it change ? Will it work ?

**Reply by davidpdrsn** [reply]
Middleware can't accesss state set by `.with_state()`. You have to pass it directly to something like `from_fn_with_state`. Hoping to fix that one day but dunno how to do it.

**Reply by oscar6echo** [reply]
Ok, this adds clarity.  
I rephrase and summarize to make sure I get it - please correct me if wrong:  

Each middleware is only aware of the state it receives in `from_fn_with_state()`. So each mw can use the same or different, presumably more specific, states. And theses are distinct from the one passed in `with_state()`.  The latter is potentially available to handlers only. 
Finally the way for a middleware to pass a state to handlers is via an `Extension` (not ideal as not type safe) or via a state if this one is passed *both* to `from_fn_with_state()` and `with_state()`.  

Is that a valid description ?

**Reply by davidpdrsn** [reply]
> Each middleware is only aware of the state it receives in from_fn_with_state()

Yes they are distinct.

> Finally the way for a middleware to pass a state to handlers is via an Extension

https://docs.rs/axum/latest/axum/middleware/index.html#passing-state-from-middleware-to-handlers

You cannot use state for this because state is global within a router. For request specific state use extensions.

**Reply by oscar6echo** [reply]
Ok clear thx :pray: 

> Middleware can't accesss state set by .with_state(). You have to pass it directly to something like from_fn_with_state. Hoping to fix that one day but dunno how to do it.

Btw I do not dislike this API: it allows fine grained control.  
Various ways to pass states around are rather good I'd say, as long as it clearly documented.

**Reply by jasondemorrow** [reply]
@thomasmost Thanks for the for the well written question and follow-ups, I've learned a lot from this thread. You mentioned being interested in alternative implementations for your router above. I'm using `axum-keycloak-auth` for my auth layer, and its extension allows defining roles as an [enum](https://github.com/lpotthast/axum-keycloak-auth/blob/5c3149474b398b475a9e2f7d1f1c8a8e0f72bb74/README.md?plain=1#L29).

This simplifies the router definition BUT it has the disadvantage of parsing roles directly from the JWT payload. It could be that your middleware is fetching roles from a DB, which is a better practice. If that's the case, you might include in your app state a client or function for verifying roles against a given authenticated user:

```
pub enum Role { ... }

impl AppState {
  pub async fn verify_roles(roles: Vec<Role>, user: SessionUser) -> bool { ... }
}
```

The advantage of this approach is that it doesn't rely on layer ordering. The disadvantage is that it requires some boilerplate per handler. You could reduce this with something like:

```
pub async fn require_roles(roles: Vec<Role>, user: SessionUser) -> Result<(), MissingRolesError> { ... }
```

I'd prefer an approach like this, but I don't think one is clearly better than the other.
